pkg/ipam/node.go

```diff
@@ -782,3 +748,51 @@ func (n *Node) syncToAPIServer() (err error) {
 
 	return err
 }
+
+// update is a helper function for syncToAPIServer(). This function updates the
+// CiliumNode resource spec or status depending on `status`. The resource is
+// updated from `origNode` to `node`.
+func (n *Node) update(
+	origNode, node *v2.CiliumNode,
+	attempts int,
+	status bool,
+) (*v2.CiliumNode, *v2.CiliumNode, error) {
+	scopedLog := n.logger()
+
+	var (
+		updatedNode    *v2.CiliumNode
+		updateErr, err error
+	)
+
+	if status {
+		updatedNode, updateErr = n.manager.k8sAPI.UpdateStatus(origNode, node)
+	} else {
+		updatedNode, updateErr = n.manager.k8sAPI.Update(origNode, node)
+	}
+
+	if updatedNode != nil && updatedNode.Name != "" {
+		node = updatedNode.DeepCopy()
+		if updateErr == nil {
+			return origNode, node, nil
+		}
+	} else if updateErr != nil {
+		scopedLog.WithError(updateErr).WithFields(logrus.Fields{
+			logfields.Attempt: attempts,
+		}).Warning("Failed to update CiliumNode spec")
+
+		node, err = n.manager.k8sAPI.Get(node.Name)
+		if err != nil {
+			return origNode, node, err
+		}
+
+		// Propagate the error in case we exit the loop without succeeding in
+		// updating the resource.
+		err = updateErr
+		node = node.DeepCopy()
+		origNode = node.DeepCopy()
```

Comment:
Can you spend a couple cycles to understand why are we DeepCopying this? I've seen like 6 DeepCopies already and, since the node structure is coming from kube-apiserver directly, there is no need to perform these many DeepCopies.

Reply:
I think the reason we have deepcopy in this specific case is because if we successfully update the CiliumNode status, then we want the reference to `origNode` and `node` to be updated for the next call to update the CiliumNode spec. I'll add a comment to clarify propagating this, similar to the error.

Reply:
Synced offline. Will make a followup PR to remove the deep copies and just update the pointers by dereferencing, while also removing the redundant return values.